### DISCUSSION TITLE
pnpm workspace package resolution, how it works ?

### DISCUSSION THREAD

**Discussion by billouboq** [prisma/prisma#20149] [Q&A] [answered]
Hi,

I was using yarn version 1 workspace.

I have 4 app package and 4 common packages. One of the common package export some tooling script and config from webpack.
When I was using yarn workspace, I just put my tooling package inside the dev-deps of all my app package and everything worked fine.

When I switched to pnpm, I had to put all the dependencies of my "tooling" package inside each package otherwise when I was starting webpack it was saying "module not found". Is this the normal behaviour ? Am I missing something ?

I don't understand why it worked with yarn but pnpm ask me to put ALL dev dependancies that my tooling package have inside all the places where I use it, feels unintuitive and makes it less appealing to create packages. 

Here is my tooling package package.json (removed private properties in it) :

```json
{
	"name": "front-tooling",
	"dependencies": {
		"@babel/core": "^7.21.8",
		"@babel/plugin-proposal-class-properties": "^7.18.6",
		"@babel/plugin-proposal-export-default-from": "^7.18.10",
		"@babel/plugin-proposal-object-rest-spread": "^7.20.7",
		"@babel/plugin-proposal-optional-chaining": "^7.21.0",
		"@babel/plugin-transform-modules-commonjs": "^7.22.5",
		"@babel/plugin-transform-flow-strip-types": "^7.21.0",
		"@babel/plugin-transform-runtime": "^7.21.4",
		"@babel/preset-env": "^7.21.5",
		"@babel/preset-react": "^7.18.6",
		"@babel/preset-typescript": "^7.21.5",
		"@babel/runtime": "^7.21.5",
		"@pmmmwh/react-refresh-webpack-plugin": "^0.5.10",
		"@sentry/webpack-plugin": "1.20.1",
		"babel-loader": "^9.1.2",
		"chalk": "^4.1.0",
		"core-js": "^3.30.2",
		"css-loader": "^6.7.4",
		"css-minimizer-webpack-plugin": "^4.2.2",
		"esbuild": "0.17.19",
		"file-loader": "^6.2.0",
		"html-webpack-plugin": "^5.5.1",
		"mini-css-extract-plugin": "^2.7.6",
		"react-dev-utils": "^11.0.4",
		"react-refresh": "^0.14.0",
		"resize-observer-polyfill": "^1.5.1",
		"sass": "1.62.1",
		"sass-loader": "^13.2.2",
		"terser-webpack-plugin": "^5.3.9",
		"url-loader": "^4.1.1",
		"webpack": "^5.83.1",
		"webpack-bundle-analyzer": "^4.8.0",
		"webpack-dev-server": "^4.15.0",
		"webpackbar": "^5.0.2",
		"buffer": "^6.0.3",
		"process": "^0.11.10"
	}
}
```

And here is how I import it where I use it :

```json
{
	"name": "back-office",
	"devDependencies": {
		"front-tooling": "workspace:^",
		"@types/react": "^16.9.56",
		"@types/react-dom": "^16.9.9",
		"@types/react-redux": "^7.1.11",
		"@types/react-select": "^3.0.26",
		"@types/react-toggle": "^4.0.2"
	}
}
```

With pnpm workspace being very simple :

```yaml
packages:
  # include packages in subfolders (e.g. apps/ and packages/)
  - 'packages/**'
```

**Comment by ludralph** [comment, marked as answer]
Hi @billouboq  👋 

The behavior you described with `pnpm` is the expected behavior. When using `pnpm` workspaces, each package maintains its own dependencies separately. This means that if you have a common package that exports some tooling scripts and webpack configurations, you need to include all the dependencies required by that tooling package in each package where you use it. This behavior ensures that each package has its own isolated set of dependencies and can be built and run independently. It helps avoid conflicts between different packages and provides better encapsulation.

On the other hand, with `Yarn` workspaces, all the packages share the same node_modules folder at the workspace root. This means that if a package depends on another package, it can resolve the dependencies from the shared node_modules folder, simplifying the configuration. The difference in behavior between `pnpm` and `Yarn` in this case is due to their different approaches to managing dependencies in a workspace environment. While `Yarn` uses a shared node_modules folder, `pnpm` isolates dependencies for each package.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
